### ISSUE TITLE
'noImplicitReturns' doesn't report an error if any branch returns 'any'

### ISSUE THREAD

**Issue by DanielRosenwasser** [Microsoft/TypeScript#13112]
The following correctly reports an error on `foo` under `noImplicitReturns`.

```ts
function foo(y: number, z: number) {
    if (y)
        return z;
}
```

However, removing the type annotations from `y` and `z` squelches the error.

```ts
function foo(y, z) {
    if (y)
        return z;
}
```

**Comment by vladima** [comment]
Discussed offline with @mhegazy and @billti - current behavior for .ts files is by design - we don't report `noImplicitReturns` error if inferred return type is `any` or `void`. It makes sense however to report it for `.js` files and we already have an [issue](https://github.com/Microsoft/TypeScript/issues/6802) to track this. Closing this one in favor of it.
